### PR TITLE
gossip: remove frequent gossiping of gossip client connections

### PR DESCRIPTION
**Pull request by a-robinson**
These gossip-clients keys make up two thirds or more of the gossip traffic in various large clusters I've inspected, consuming almost an entire CPU core in the worst case I've seen. They don't provide enough value to justify that sort of ongoing cost, so this commit removes them entirely as well as the periodic logging of the gossip network and the crdb_internal.gossip_network table, both of which relied on them.

These gossip-clients keys make up two thirds or more of the gossip
traffic in various large clusters I've inspected, consuming almost an
entire CPU core in the worst case I've seen. They don't provide enough
value to justify that sort of ongoing cost, so this commit removes them
entirely as well as the periodic logging of the gossip network and the
crdb_internal.gossip_network table, both of which relied on them.

Release note (backward-incompatible change): We've stopped
supporting/populating the crdb_internal.gossip_network table. It was an
internal table with no API guarantees (so perhaps no meriting a release
note?).

Release note (performance improvement): Significantly reduced CPU usage
of the underlying gossip network in large clusters.

---

Informs #51838 (largely fixes it for practical purposes, although there's likely still more that could be done)

This is clearly going to break [the gossip roachtest](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/tests/gossip.go#L50-L58), but between @irfansharif kindly volunteering to fix that up separately and his existing TODO in that file I've left that out of this change.

I don't know if completely removing the gossip_network table is really the best idea or if it should just be left in and only populated with the clients from the local node. For example, when run in a mixed version cluster does `debug zip` run all of its sql commands against the local node or does it run some against remote nodes? If an old node ever tries to query the `gossip_network` table on a different node it could have a bad time.

@irfansharif @kvoli

### CODE REVIEW

**Comment by a-robinson** [comment]
(apologies in advance if some tests fail in CI -- my laptop is able to build cockroach but not really set up to run the full test suite, so I'm relying on CI to point out any issues)

File: pkg/sql/crdb_internal.go
Comment (kvoli) [inline pkg/sql/crdb_internal.go:3942]: Was this change needed to pass some tests?

**Review by kvoli — COMMENTED** [review]
Thanks for the PR @a-robinson.

> For example, when run in a mixed version cluster does debug zip run all of its sql commands against the local node or does it run some against remote nodes?

I'm not certain on this one. I think local node but I'm not familiar here. cc @irfansharif.


The failing test `TestGossipPropagation` seems due to no component actively removing expired infos. Previously the `clientsTimer` would tick in this test every 2 seconds so it would eventually succeed.

I think in a real cluster, given the store gossip @ 10s so there is still something removing it, should be fine - just specific to this test. Triggering gossip w/ a key inside the retry loop should fix

```
mustAdd(remote, "remote-trigger", nil, 2 * time.Minute)
```

**Comment by a-robinson** [comment]
> I think in a real cluster, given the store gossip @ 10s so there is still something removing it, should be fine - just specific to this test. Triggering gossip w/ a key inside the retry loop should fix

Thanks, I suspect so too. I just haven't had a good chance to fix it up yet, but will do soon.

File: pkg/sql/crdb_internal.go
Comment (a-robinson) [reply pkg/sql/crdb_internal.go:3942]: I haven't actually tried running any tests without this. It just seemed like it would be pretty bad if `crdb_internal.gossip_nodes` suddenly switched to always saying that all nodes are not live (due to the gossip clients keys being missing). Especially since IIRC it's the table that's used by CLI commands like `cockroach node status`.

And basing it on node liveness seems like a reasonable alternative, although I'm open to other ideas.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89613)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @a-robinson, @irfansharif, and @kvoli)

___
*[`pkg/sql/crdb_internal.go` line 4234 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89613#-NE7B7NIACIuZ-2K4xhh:-NE7B7NIACIuZ-2K4xhi:b-7bdwv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e06d2286b011096526eda7f2d7f7bb7acea0ae84/pkg/sql/crdb_internal.go#L4234)):*
> ```Go
> // crdbInternalGossipNetwork exposes the local view of the gossip network (i.e
> // the gossip client connections from source_id node to target_id node).
> var crdbInternalGossipNetworkTable = virtualSchemaTable{
> ```

drive-by comment: i'd be slightly more inclined to make this an empty table (or do the idea to only show local clients) rather than removing the table outright

we could also have it do `p.BufferClientNotice` to warn that this table is no longer reliable

**Comment by irfansharif** [comment]
I'll just take over this PR + fallout, through next week.

**Comment by a-robinson** [comment]
> I'll just take over this PR + fallout, through next week.

Thanks!

**Comment by a-robinson** [comment]
Congrats on getting 22.2 out! I'd hope this is something that will make it into v23.1?

Let me know if it'd help for me to clean something up here, although it's unlikely that I'm the best person to make the call on what to do with the crdb_internal.gossip_network` table (or any other compatibility concerns) at this point.

**Review by irfansharif — COMMENTED** [review]
Sorry that this slipped, I was over optimistic with how many hours I had in a week. Rebased to let another run of CI through to see what shakes out. Shall get rid of the stale roachtests after, and try Rafi's idea to make it an empty table with a warning. I'm personally aiming for this to be backported to 22.2 to not have to wait for another 6 months. We're breaking compatibility of an internal table, but it's an internal table. If folks complain (I doubt), I'll re-add under some disabled-by-default flag.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89613#-:-NJfHl8D10O3Ii7CZBxS:bgat73x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @a-robinson, @herkolategan, @kvoli, and @smg260)

**Comment by a-robinson** [comment]
Sounds good, thanks!

**Review by irfansharif — COMMENTED** [review]
I need a reviewer here. @kvoli and @tbg, up for it? Alex did most of the work so you know it's good. Do review it with backportability to 22.2 (at the very least) in mind. Was thinking we'd let this bake for a two-ish weeks pre-backport. If it doesn't feel eligible for a backport now that 22.2.0 was published publicly, I'll be a bit regretful of dropping it earlier but would understand.

---

> I don't know if completely removing the gossip_network table is really the best idea or if it should just be left in and only populated with the clients from the local node. 

I've re-added the table but left it empty. It just makes it slightly more palatable for a backport. 

> we could also have it do p.BufferClientNotice to warn that this table is no longer reliable

Done.

> For example, when run in a mixed version cluster does debug zip run all of its sql commands against the local node or does it run some against remote nodes? If an old node ever tries to query the gossip_network table on a different node it could have a bad time.

Runs it against the local node. But we do want to support older version CRDB binaries in a point release querying newer point release CRDB servers, so keeping the table around seems ok. For mixed-version clusters while the cluster version hasn't yet been finalized, we want to be able to use the older binary generally. There too keeping the table around seems helpful. In a future release we can get rid of it entirely.

> The failing test TestGossipPropagation seems due to no component actively removing expired infos. 

Fixed.

> This is clearly going to break the gossip roachtest, but between @irfansharif kindly volunteering to fix that up separately and his existing TODO in that file I've left that out of this change.

Fixed.

>  It just seemed like it would be pretty bad if crdb_internal.gossip_nodes suddenly switched to always saying that all nodes are not live (due to the gossip clients keys being missing). Especially since IIRC it's the table that's used by CLI commands like cockroach node status. And basing it on node liveness seems like a reasonable alternative, although I'm open to other ideas.

Keeping gossip_nodes backed by the gossiped view of node liveness seems better than using the KV node liveness. It's generally good that these tables are highly available and don't need leaseholder participation. We could get that with liveness too by looking at the in-memory state on the node, but I'll keep it as is.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89613#-:-NJgPfLzC4MjIHlT8inl:b3aj7mi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @a-robinson and @kvoli)

**Review by tbg — APPROVED** [review]
:lgtm: thanks for pushing this over the finish line!

> Reviewed 5 of 22 files at r1, 8 of 17 files at r2, 1 of 2 files at r3, 3 of 11 files at r4, all commit messages.
> Dismissed @kvoli from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89613#-:-NJifc3W9mYZZTl8Hu8-:bcw1pe7)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @a-robinson and @irfansharif)

___
*[`-- commits` line 9 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89613#-NJidoZq1CbJP5pPKffE):*
Still reading the PR, but is there a chance in the `./cockroach node status` data sources (see cli roachtest change)? If so, worth pointing out.
___
*[`pkg/cmd/roachtest/tests/cli.go` line 112 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89613#-NJidhFcDAIhokj4bu8y:-NJidhFcDAIhokj4bu8z:b-gylpr4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2114678be6d1ac279ec1fed3420fb635cc255476/pkg/cmd/roachtest/tests/cli.go#L112)):*
> ```Go
> 	waitUntil([]string{
> 		"is_available is_live",
> 		"false false",
> ```

I am confused by this change, what's going on here?
___
*[`pkg/sql/crdb_internal.go` line 4108 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89613#-NJif3Wk6NVHwPz_ehsJ:-NJif3Wk6NVHwPz_ehsK:bqhw8l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2114678be6d1ac279ec1fed3420fb635cc255476/pkg/sql/crdb_internal.go#L4108)):*
> ```Go
> 			if err := g.GetInfoProto(gossip.MakeNodeLivenessKey(d.NodeID), &gossipLiveness); err == nil {
> 				if now.Before(gossipLiveness.Expiration.ToTimestamp().GoTime()) {
> 					alive[d.NodeID] = true
> ```

This works, but I wonder if this is the canonical way to do this sort of thing in unified architecture. I would've thought the tenant would have a liveness view of KV around somewhere (possibly powered by Gossip under the hood). Not to block the merge - after all this works - but maybe @knz can take a look and see if this is palatable long-term.

**Review by kvoli — APPROVED** [review]
> Do review it with backportability to 22.2 (at the very least) in mind. Was thinking we'd let this bake for a two-ish weeks pre-backport. If it doesn't feel eligible for a backport now that 22.2.0 was published publicly, I'll be a bit regretful of dropping it earlier but would understand.

I don't see an issue backporting.

:lgtm: 

> Reviewed 2 of 22 files at r1, 1 of 17 files at r2, 8 of 11 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89613#-:-NJk-zYv3Ac0UnE9EqOt:bwrutne)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @a-robinson and @irfansharif)

**Review by irfansharif — COMMENTED** [review]
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89613#-:-NJpH-AS7J4ykzj6f2LT:b-oduykh)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @a-robinson and @tbg)

___
*[`-- commits` line 9 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89613#-NJidoZq1CbJP5pPKffE):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Still reading the PR, but is there a chance in the `./cockroach node status` data sources (see cli roachtest change)? If so, worth pointing out.
</blockquote></details>

There's no change.
___
*[`pkg/cmd/roachtest/tests/cli.go` line 112 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89613#-NJidhFcDAIhokj4bu8y:-NJpFODU5xD9s-3lnG1p:b-6opk1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/2114678be6d1ac279ec1fed3420fb635cc255476/pkg/cmd/roachtest/tests/cli.go#L112)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I am confused by this change, what's going on here?
</blockquote></details>

It's because of the changes under pkg/sql/crdb_internal.go. Before the is_live column was computed off of the existence of a gossip clients key, which for n1 in test test, would've existed (n1 isn't taken down). But now that we no longer have these keys, is_live is backed by gossiped back liveness. And n1 can't heartbeat its liveness because we've killed 2/3 nodes.
___
*[`pkg/sql/crdb_internal.go` line 4108 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89613#-NJif3Wk6NVHwPz_ehsJ:-NJpGK5o1zcBE1kFmV6d:bweiesv) ([raw file](https://github.com/cockroachdb/cockroach/blob/2114678be6d1ac279ec1fed3420fb635cc255476/pkg/sql/crdb_internal.go#L4108)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This works, but I wonder if this is the canonical way to do this sort of thing in unified architecture. I would've thought the tenant would have a liveness view of KV around somewhere (possibly powered by Gossip under the hood). Not to block the merge - after all this works - but maybe @knz can take a look and see if this is palatable long-term.
</blockquote></details>

I'll defer to Rafa to sort out what a tenant-with-capabilities view of this would look like. I assume it's just going to be a table populated by data returned through the Connector interface, which in turn would do a very similar thing to what we're doing here.

**Review by a-robinson — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89613)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)

___
*[`pkg/cmd/roachtest/tests/cli.go` line 112 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89613#-NJidhFcDAIhokj4bu8y:-NJpJTR47TIW1K3Ym6Gq:b-h6444h) ([raw file](https://github.com/cockroachdb/cockroach/blob/2114678be6d1ac279ec1fed3420fb635cc255476/pkg/cmd/roachtest/tests/cli.go#L112)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

It's because of the changes under pkg/sql/crdb_internal.go. Before the is_live column was computed off of the existence of a gossip clients key, which for n1 in test test, would've existed (n1 isn't taken down). But now that we no longer have these keys, is_live is backed by gossiped back liveness. And n1 can't heartbeat its liveness because we've killed 2/3 nodes.
</blockquote></details>

FWIW it may be possible to use gossiped store descriptors instead of gossiped liveness since my understanding is that those get refreshed every couple minutes, but I don't know whether that would really be better overall. Using liveness seemed better to me at the time, but I didn't consider what the output should look like during cluster instability.

**Comment by irfansharif** [comment]
blathers backport 22.2

**Comment by erikgrinaker** [comment]
I think the 22.2 backport here broke backwards compatibility with 22.1. In a mixed 22.2/22.1 cluster, running `cockroach node status` on a 22.1 node will report `is_live = false` for 22.2 nodes. I'll have a closer look at exactly why.

**Comment by erikgrinaker** [comment]
Ah, right, I misread the keys. 22.1 uses `GossipClientsKey` to determine liveness, but 22.2 no longer gossips that key and relies on `NodeLivenessKey` instead. This should've been behind a version gate.

I don't know if there's much we can do about this now. 22.1 is out of support with no further releases planned. We can consider backporting a retroactive 22.2 version gate, but it's not clear that it's worth the risk and effort.

**Comment by erikgrinaker** [comment]
I guess we'll have the same problem with mixed 22.2 patch releases before/after this backport, and with mixed 22.2/23.1 clusters before this backport. The right approach here would've been to add a 23.1 version gate and not backport to 22.2.

I think the best option here would be to backport a 23.1 version gate, and to semi-revert the 22.2 backport such that it still gossips the `GossipClientsKey` but doesn't actually use it for anything. Alternatively, live with a broken `is_live` in `cockroach node status` with certain combinations of 22.2 mixed-version clusters.

**Comment by erikgrinaker** [comment]
For reference, this was initially released in v22.2.3.

**Comment by erikgrinaker** [comment]
Opened #103788 with a partial 22.2 revert. I'm not entirely convinced we want to do that, considering the cost of gossiping these, but breaking backwards compatibility is also not great.

**Comment by irfansharif** [comment]
> This should've been behind a version gate.

This was intentional BTW, since we did want to backport it to 22.2 and backports can't carry version gates. The `is_live` breakage in mixed version settings is slightly unfortunate but it's less meaningful IMO than the CPU% the backport reduces. 

> I'm not entirely convinced we want to do that, considering the cost of gossiping these

I'm with you. I feel like a release note pointing to newer patch releases is sufficient: https://github.com/cockroachdb/cockroach/pull/103788#pullrequestreview-1440106572.

**Comment by erikgrinaker** [comment]
We can't add version gates, but we can piggyback on the existing V22_2 version gate.

**Pull request merged**
